### ISSUE TITLE
There is need of more complex math calculations like sine and cosine

### ISSUE THREAD

**Issue by ibrahimcengiz** [dbaldwin/DroneBlocks-Web#20]
There is a photo technique as 'flatland'. (http://www.thisiscolossal.com/2017/03/flatland-ii-landscapes-aydin-buyuktas/)

i tried to automate the proccess with droneblocks but the need of sine and cosine functions are inevitable.

**Comment by dbaldwin** [comment]
I'm very familiar with this technique and have not been able to reproduce it successfully. We've discussed internally how to do a tutorial lesson of this technique and maybe you can help.

Can you share how sine and cosine would help? If you could explain the process in the following format:

- takeoff to 100
- pitch gimbal to 0 degrees
- take photo
- yaw aircraft right 60 degrees

We can look into this. Actually, implementing sine and cosine would be really easy. We just need some example scripts that we can test. Let me know and we can consider this for the next DroneBlocks release. We can also get you on the beta to help test this method.

Thanks for the feedback.

**Comment by dbaldwin** [comment]
@ibrahimcengiz I'm posting your great overview of the math to pull this off.... thanks for putting this together. We will get to work on it asap!

[Flatland_DroneBlocks (1).pdf](https://github.com/dbaldwin/DroneBlocks-Web/files/1095513/Flatland_DroneBlocks.1.pdf)

**Comment by kbquist** [comment]
@ibrahimcengiz 

This is amazing work.  I'll try to add the logic for the "fly forward" part of the program.

we would add a variable like "XPosition" Which would start out equal to the MaxAltitude (I've kept the signs positive vs. the negative in your example)

//this variable will be used to calculate how far forward we need to fly on each step
XPosition = XPosition - MaxAltitude * cos((90/NumberOfPhotos-1)*Sqn)

Then, we would have a Fly Forward XPosition distance block

Dennis - we'll also need to make sure we're working with degrees and not radians.

**Comment by dbaldwin** [comment]
I've been experimenting with this in JavaScript because that's what we're using to translate block code to something usable with the DJI SDK. So the JS code looks like this:

```
function rotate(x, y, angle) {
  var newx = x * Math.cos(toRadians(angle)) - y * Math.sin(toRadians(angle));
  var newy = y * Math.cos(toRadians(angle)) + x * Math.sin(toRadians(angle));
  return [Math.round(newx), Math.round(newy)];
}

function toRadians(angle) {
  return angle * (Math.PI/180);
}
```

Using these functions with the following inputs:

```
rotate(120, 0, 0);

// Yields [120, 0]
```

```
rotate(120, 0, 30);

// Yields [104, 60]
```

```
rotate(120, 0, 60);

// Yields [60, 104]
```

```
rotate(120, 0, 90);

// Yields [0, 120]
```

Just wanted to document this here as we build out the ability to support sine and cosine. We obviously need to be able to support variables in loop and pitch gimbal blocks.

**Comment by dbaldwin** [comment]
@ibrahimcengiz @kbquist when you get a chance will you try to construct a flatland mission on our staging server? We've added the following to this build:

- Variable input for take off block
- Variable input for loop block
- Variable input for pitch gimbal block
- Math sine and cosine functions

Keep in mind that this is not yet linked up to the SDK. We'll need to push a new beta release of DroneBlocks soon. But if you can get away with constructing a mission using the new features above we can get a beta out next week. Here is the staging server link:

https://droneblocks-staging.firebaseapp.com/?n=1

Please share your mission when you have it ready. We'd love to see it.

**Comment by ibrahimcengiz** [comment]
@dbaldwin @kbquist 
i am working on it. i will share my algorithm as soon as it is ready.
i also want to verify my algorithm using python.

**Comment by ibrahimcengiz** [comment]
I have done my algorithm, it looks good. Could you check it too?
thanks

ipad : droneblocks://?missionId=-KnQRdACilq0196XgVNl
desktop : https://dev.droneblocks.io?share=1&missionId=-KnQRdACilq0196XgVNl

Python script
[Flatland.zip](https://github.com/dbaldwin/DroneBlocks-Web/files/1099960/Flatland.zip)
![flatland_blocks](https://user-images.githubusercontent.com/8030006/27511246-7e2387e4-5928-11e7-98d8-35c3a8c18e12.png)

**Comment by dbaldwin** [comment]
Here is the link to the mission. Our staging code references a production url so the share link you posted is broken.

https://droneblocks-staging.firebaseapp.com/?share=1&missionId=-KnQRdACilq0196XgVNl

A couple of thoughts regarding your code. Should we have a take photo block before the loop since that will be the straight forward shot?

Looking at your DroneBlocks mission and also running your Python code I think sqn needs to be initialized to 1. Otherwise in the first loop the altitude will be 0. If we initialize it to 1 then your Python script yields the following.

```
Fly Forward to : 4.06235034732
Fly to Altitude : 51.6202338128
Pitch gimbal to : 9
take photo
Fly Forward to : 12.0870347368
Fly to Altitude : 101.969561718
Pitch gimbal to : 18
take photo
Fly Forward to : 19.8141326429
Fly to Altitude : 149.808367899
Pitch gimbal to : 27
take photo
Fly Forward to : 27.0534005574
Fly to Altitude : 193.958846351
Pitch gimbal to : 36
take photo
Fly Forward to : 33.6266054952
Fly to Altitude : 233.333998821
Pitch gimbal to : 45
take photo
Fly Forward to : 39.371913145
Fly to Altitude : 266.964397043
Pitch gimbal to : 54
take photo
Fly Forward to : 44.1478722795
Fly to Altitude : 294.022050349
Pitch gimbal to : 63
take photo
Fly Forward to : 47.8368973273
Fly to Altitude : 313.840791059
Pitch gimbal to : 72
take photo
Fly Forward to : 50.3481633662
Fly to Altitude : 325.932675729
Pitch gimbal to : 81
take photo
Fly Forward to : 51.6198422601
Fly to Altitude : 329.999998469
Pitch gimbal to : 90
take photo
Fly Forward to : 51.6206248865
Fly to Altitude : 325.942620552
Pitch gimbal to : 99
take photo
```
That's also why I proposed taking a photo before the loop block. Just thinking about loud and this looks really good.

**Comment by dbaldwin** [comment]
Oh yeah.... a couple other things I just noted. It looks like there is one extra loop. The camera obviously can't pitch to 99 degrees so I think the repeat block should just be set to "NumberOfPhotos" and not "NumberOfPhotos + 1".

The other thing is that the gimbal pitch is a negative value. So the 90 multiplier should be -90. Here is the revised Python code with all the changes I've mentioned above:

```
import math

MaxAlt = 330
NumOfPht = 10
Sqn = 1
XPosition = 0

for i in range(NumOfPht):
    NewXPosition = MaxAlt-MaxAlt*math.cos(Sqn*1.5707/NumOfPht)
    flyfwrd = NewXPosition-XPosition
    print ('Fly Forward to : ' + str(flyfwrd))
    FlytoAltitude = MaxAlt * math.sin(Sqn*1.5707/NumOfPht)
    print ('Fly to Altitude : ' + str(FlytoAltitude))
    XPosition = NewXPosition
    gimbalangle = Sqn * (-90/NumOfPht)
    print ('Pitch gimbal to : ' + str(gimbalangle))
    print ('take photo')
    Sqn = Sqn + 1
```

**Comment by ibrahimcengiz** [comment]
_A couple of thoughts regarding your code. Should we have a take photo block before the loop since that will be the straight forward shot?_

- Yes, first photo should be parallel to the ground, i saw some people took their photos in first shot.

_Otherwise in the first loop the altitude will be 0_

- i don't think so, in the beginning of code we take off to altitude 10 ft.

Sqn should be start at 0 imo but this makes first loop ineffective. Does fly forward 0 feet command cause an problem?

**Comment by dbaldwin** [comment]
Currently we cannot implement a fly forward block with 0 feet. The SDK requires that a waypoint be at least 1 m from the current location. The bigger problem is the 0 altitude block. The current block implementation is absolute. You can see that it's titled "change altitude to xxxx" which means it's looking for an absolute value. Setting it to 0 would likely produce an error in the SDK but if not it would actually descend to the ground. So for now we should just skip this loop or start with the next iteration.

**Comment by ibrahimcengiz** [comment]
Yes, you are right, my mistake
i have updated the code, i added an if block for disabling the first loop.
I added the take off altitude parameter but as i understand every program have to be start with take off block. However i want to assign takeoffaltitude parameter to a value before take off. Anyway this is not a big issue

![flatland_blocks_ 06_26](https://user-images.githubusercontent.com/8030006/27525657-6b0623d0-5a48-11e7-86fd-f90d4688cfda.png)


https://dev.droneblocks.io?share=1&missionId=-KnXbvm8QbML2Wt4kUfC

**Comment by dbaldwin** [comment]
@ibrahimcengiz we removed the restriction for the takeoff block a couple versions ago. Now you can fly to a location and begin a mission mid-air. So the takeoff block isn't necessary.

@kbquist since we are testing on the staging server the shared mission is located here:

https://droneblocks-staging.firebaseapp.com/?share=1&missionId=-KnXbvm8QbML2Wt4kUfC#

This is a small bug in our staging vs production code.

**Comment by dbaldwin** [comment]
BTW @ibrahimcengiz your block code looks good to me. I think we should get a beta out soon and do some testing. What do you think?

**Comment by hdrpano** [comment]
I have created the loop function as I use it for my inception pictures

![dp-i-p](https://user-images.githubusercontent.com/29736313/27602022-9c50c416-5b71-11e7-9c8c-72152ba284d6.JPG)

See this illustration

![dp-i-p 2](https://user-images.githubusercontent.com/29736313/27602172-26450a6a-5b72-11e7-807b-c15b8f1a7a52.JPG)

**Comment by ibrahimcengiz** [comment]
Yes we must test this huge update :)

Btw i want to ask an off topic question.
Can we do 2 command simultaneously? For example fly forward and pitch gimbal in the same time.

**Comment by dbaldwin** [comment]
@ibrahimcengiz It's not currently possible to run two commands in parallel. We'll need to look into this in the future. If you have a specific use-case please create a new task and describe what you'd like to do.
